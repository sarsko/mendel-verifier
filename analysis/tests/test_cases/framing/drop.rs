use std::sync::{Mutex, MutexGuard};

fn unknown<T>(x: T) {}

#[analyzer::run]
fn test(guard2: MutexGuard<i32>) {
    let mutex = Mutex::new(123);
    let mut guard = mutex.lock().unwrap();
    guard = guard2;
}

fn main() {}
