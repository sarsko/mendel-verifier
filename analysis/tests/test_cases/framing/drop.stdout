Analyzing file $DIR/drop.rs using FramingAnalysis...
Result for function test():
{
  "bb0": [
    [
      [
        "state:",
        {
          "frame_accessible": [
            "_1"
          ],
          "frame_locally_shared": [
            "_1"
          ],
          "frame_owned": [
            "_1"
          ]
        },
        "statement: StorageLive(_2)"
      ]
    ],
    "state before terminator:",
    {
      "frame_accessible": [
        "_1"
      ],
      "frame_locally_shared": [
        "_1"
      ],
      "frame_owned": [
        "_1"
      ]
    },
    "terminator: _2 = std::sync::Mutex::<i32>::new(const 123_i32) -> [return: bb1, unwind: bb11]",
    {
      "bb1": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ],
      "bb11": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb1": [
    [
      [
        "state:",
        {
          "frame_accessible": [
            "_1",
            "_2"
          ],
          "frame_locally_shared": [
            "_1",
            "_2"
          ],
          "frame_owned": [
            "_1"
          ]
        },
        "statement: FakeRead(ForLet(None), _2)"
      ],
      [
        "state:",
        {
          "frame_accessible": [
            "_1",
            "_2"
          ],
          "frame_locally_shared": [
            "_1",
            "_2"
          ],
          "frame_owned": [
            "_1",
            "_2"
          ]
        },
        "statement: StorageLive(_3)"
      ],
      [
        "state:",
        {
          "frame_accessible": [
            "_1",
            "_2"
          ],
          "frame_locally_shared": [
            "_1",
            "_2"
          ],
          "frame_owned": [
            "_1",
            "_2"
          ]
        },
        "statement: StorageLive(_4)"
      ],
      [
        "state:",
        {
          "frame_accessible": [
            "_1",
            "_2"
          ],
          "frame_locally_shared": [
            "_1",
            "_2"
          ],
          "frame_owned": [
            "_1",
            "_2"
          ]
        },
        "statement: StorageLive(_5)"
      ],
      [
        "state:",
        {
          "frame_accessible": [
            "_1",
            "_2"
          ],
          "frame_locally_shared": [
            "_1",
            "_2"
          ],
          "frame_owned": [
            "_1"
          ]
        },
        "statement: _5 = &_2"
      ]
    ],
    "state before terminator:",
    {
      "frame_accessible": [
        "_1",
        "_2"
      ],
      "frame_locally_shared": [
        "_1"
      ],
      "frame_owned": [
        "_1"
      ]
    },
    "terminator: _4 = std::sync::Mutex::<i32>::lock(move _5) -> [return: bb2, unwind: bb11]",
    {
      "bb11": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ],
      "bb2": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb2": [
    [
      [
        "state:",
        {
          "frame_accessible": [
            "_1",
            "_2",
            "_4"
          ],
          "frame_locally_shared": [
            "_1",
            "_4"
          ],
          "frame_owned": [
            "_1",
            "_4"
          ]
        },
        "statement: StorageDead(_5)"
      ]
    ],
    "state before terminator:",
    {
      "frame_accessible": [
        "_1",
        "_2"
      ],
      "frame_locally_shared": [
        "_1"
      ],
      "frame_owned": [
        "_1"
      ]
    },
    "terminator: _3 = std::result::Result::<std::sync::MutexGuard<'_, i32>, std::sync::PoisonError<std::sync::MutexGuard<'_, i32>>>::unwrap(move _4) -> [return: bb3, unwind: bb10]",
    {
      "bb10": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ],
      "bb3": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb3": [
    [
      [
        "state:",
        {
          "frame_accessible": [
            "_1",
            "_2",
            "_3"
          ],
          "frame_locally_shared": [
            "_1",
            "_3"
          ],
          "frame_owned": [
            "_1",
            "_3"
          ]
        },
        "statement: StorageDead(_4)"
      ],
      [
        "state:",
        {
          "frame_accessible": [
            "_1",
            "_2",
            "_3"
          ],
          "frame_locally_shared": [
            "_1",
            "_3"
          ],
          "frame_owned": [
            "_1"
          ]
        },
        "statement: FakeRead(ForLet(None), _3)"
      ],
      [
        "state:",
        {
          "frame_accessible": [
            "_1",
            "_2"
          ],
          "frame_locally_shared": [
            "_1"
          ],
          "frame_owned": [
            "_1"
          ]
        },
        "statement: StorageLive(_6)"
      ],
      [
        "state:",
        {
          "frame_accessible": [
            "_2"
          ],
          "frame_locally_shared": [],
          "frame_owned": []
        },
        "statement: _6 = move _1"
      ]
    ],
    "state before terminator:",
    {
      "frame_accessible": [
        "_2"
      ],
      "frame_locally_shared": [],
      "frame_owned": []
    },
    "terminator: replace(_3 <- move _6) -> [return: bb4, unwind: bb8]",
    {
      "bb4": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ],
      "bb8": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb4": [
    [],
    "state before terminator:",
    {
      "frame_accessible": [
        "_2"
      ],
      "frame_locally_shared": [],
      "frame_owned": []
    },
    "terminator: drop(_6) -> [return: bb5, unwind: bb9]",
    {
      "bb5": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ],
      "bb9": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb5": [
    [
      [
        "state:",
        {
          "frame_accessible": [
            "_2"
          ],
          "frame_locally_shared": [],
          "frame_owned": []
        },
        "statement: StorageDead(_6)"
      ],
      [
        "state:",
        {
          "frame_accessible": [
            "_2"
          ],
          "frame_locally_shared": [],
          "frame_owned": []
        },
        "statement: _0 = const ()"
      ]
    ],
    "state before terminator:",
    {
      "frame_accessible": [
        "_0",
        "_2"
      ],
      "frame_locally_shared": [
        "_0"
      ],
      "frame_owned": [
        "_0"
      ]
    },
    "terminator: drop(_3) -> [return: bb6, unwind: bb11]",
    {
      "bb11": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ],
      "bb6": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb6": [
    [
      [
        "state:",
        {
          "frame_accessible": [
            "_0",
            "_2"
          ],
          "frame_locally_shared": [
            "_0",
            "_2"
          ],
          "frame_owned": [
            "_0",
            "_2"
          ]
        },
        "statement: StorageDead(_3)"
      ],
      [
        "state:",
        {
          "frame_accessible": [
            "_0",
            "_2"
          ],
          "frame_locally_shared": [
            "_0",
            "_2"
          ],
          "frame_owned": [
            "_0",
            "_2"
          ]
        },
        "statement: StorageDead(_2)"
      ]
    ],
    "state before terminator:",
    {
      "frame_accessible": [
        "_0"
      ],
      "frame_locally_shared": [
        "_0"
      ],
      "frame_owned": [
        "_0"
      ]
    },
    "terminator: drop(_1) -> [return: bb7, unwind: bb12]",
    {
      "bb12": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ],
      "bb7": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb7": [
    [],
    "state before terminator:",
    {
      "frame_accessible": [
        "_0"
      ],
      "frame_locally_shared": [
        "_0"
      ],
      "frame_owned": [
        "_0"
      ]
    },
    "terminator: return",
    {}
  ],
  "bb8": [
    [],
    "state before terminator:",
    {
      "frame_accessible": [],
      "frame_locally_shared": [],
      "frame_owned": []
    },
    "terminator: drop(_6) -> bb9",
    {
      "bb9": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb9": [
    [],
    "state before terminator:",
    {
      "frame_accessible": [],
      "frame_locally_shared": [],
      "frame_owned": []
    },
    "terminator: drop(_3) -> bb11",
    {
      "bb11": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb10": [
    [],
    "state before terminator:",
    {
      "frame_accessible": [],
      "frame_locally_shared": [],
      "frame_owned": []
    },
    "terminator: drop(_4) -> bb11",
    {
      "bb11": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb11": [
    [],
    "state before terminator:",
    {
      "frame_accessible": [],
      "frame_locally_shared": [],
      "frame_owned": []
    },
    "terminator: drop(_1) -> bb12",
    {
      "bb12": [
        "state:",
        {
          "frame_accessible": [],
          "frame_locally_shared": [],
          "frame_owned": []
        }
      ]
    }
  ],
  "bb12": [
    [],
    "state before terminator:",
    {
      "frame_accessible": [],
      "frame_locally_shared": [],
      "frame_owned": []
    },
    "terminator: resume",
    {}
  ]
}
